cmd/soroban-cli/src/rpc_client.rs

```diff
@@ -0,0 +1,131 @@
+use std::ops::{Deref, DerefMut};
+
+use http::{HeaderMap, HeaderName, HeaderValue};
+
+use crate::config::network::Network;
+use crate::rpc;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Rpc(#[from] rpc::Error),
+    #[error(transparent)]
+    InvalidHeaderName(#[from] http::header::InvalidHeaderName),
+    #[error(transparent)]
+    InvalidHeaderValue(#[from] http::header::InvalidHeaderValue),
+}
+
+#[derive(Debug)]
+pub struct RpcClient {
+    client: rpc::Client,
+}
+
+impl RpcClient {
+    pub fn new(network: &Network) -> Result<Self, Error> {
+        let mut additional_headers = HeaderMap::new();
+        for header in &network.rpc_headers {
+            let header_name = HeaderName::from_bytes(header.0.as_bytes())?;
+            let header_value = HeaderValue::from_str(&header.1)?;
+
+            additional_headers.insert(header_name, header_value);
+        }
+
+        let client = rpc::Client::new_with_headers(&network.rpc_url, additional_headers)?;
+        Ok(Self { client })
+    }
+}
+
+// implementing Deref in order to delegate all method calls to `rpc::Client`
+impl Deref for RpcClient {
+    type Target = rpc::Client;
+
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+// implementing DerefMut for mutable access
+impl DerefMut for RpcClient {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.client
+    }
+}
```

Comment:
The use of Deref to sub one type another one is misusing the Deref trait a bit. It is a pattern used at times, but also considered an antipattern if the benefits don't truly outweigh the cons. There's some conversation about it here: https://rust-unofficial.github.io/patterns/anti_patterns/deref.html

I don't think we need to use this pattern here. Can I suggest we instead do something a little more boring, and replace this:
```diff
-    let client = Client::new(&network.rpc_url)?;
+    let client = RpcClient::new(network)?;
```
with:
```diff
-    let client = Client::new(&network.rpc_url)?;
+    let client = network.rpc_client()?;
```
where `rpc_client` is a function on `Network` that returns an `rpc::Client` and constructs it from the necessary parameters?

Reply:
Good to know that using Deref like this is an antipattern. Thanks for the resource!

And I like this refactor - definitely simpler. I'll get a change pushed up shortly.